### PR TITLE
fs-version-002: New Versioning Mechanism

### LINKED ISSUE #797: `pbench-move-results` NG and its support structures on the server side.
* History
@tenstormavi has implemented an inotify-based implementation of the server scripts which decreases latency and will allow us to containerize the server for easier deployment in a cloud environment. However, the mechanism required a backward-incompatible change to `pbench-move-results` and the proposed mechanism to deal with it (have `pbench-move-results` call a server-side script to do what's needed) was deemed a backward step. Although this change is actually part of the current release (and is an improvement over doing things explicitly in the client), it will not be used to resolve the backward incompatibility noted above. Instead, we'll work along the lines of the following proposal (but there will be comments, so please make sure to (re-)read the whole issue before implementing anything and update the issue with any questions and problems you encounter. 

* Proposal
Based on comments by @portante, we are planning to do the following:

- Use the versioning mechanism on the server side to deal with the incompatibiliy: old clients will remain unchanged and the data will be processed exactly as they are processed today (eventually, once most clients
have been updated, we'll declare a moratorium and turn it off but no date has been set); new clients will use the new mechanism described below.

- `pbench-move-results` will copy *one* tarball per result to the server (eventually, that will become a PUT into an object filesystem). The tarball will package up a few pieces that are currently copied separately: the compresed tarball of the results, its MD5 sum and the (optional) prefix file. In addition, the controller host name will be added as additional information (currently it's encoded as part of the pathname of the tarball).

- In addition to bumping the fs version, we'll need to provide new `results-host-info` settings.

- On the server side, the inotified server will watch the download directory and provide impedance matching to the current processing pipeline (which remains substantially unchanged).

- We'll change the archive directory structure to make it deeper and not as wide, by using the MD5 string of the tarball to implement a  (on-disk) trie (like git, or gluster): e.g. the MD5 sum 7341243af6d9af1025cbec71023dbff7 will generate a pathname of `73/41/7341243af6d9af1025cbec71023dbff7` to a directory that will hold the unpacked "container" tarball. [Question: does this structure limit us in any way? Is there anything that becomes harder? Need to think about this some more]

- The (inotified) server will create a symlink to the "inner" tarball in a TODO state directory and add an entry to the list of files to be processed. From that point on, the current processing pipeline will take over and perform the same tasks it does today.

### PR DESCRIPTION
**Pull request by tenstormavi**
Fixes #797

This PR will do the following changes:

1. pbench-move-results will create a new versioning mechanism on the server side to deal with the incompatibility: old clients will remain unchanged and the data will be processed exactly as they are processed today; new clients will use the new mechanism.

 2. pbench-move-results will copy one tarball per result to the server. The tarball will package up a few pieces that are currently copied separately: the compressed tarball of the results, its MD5 sum, and the (optional) prefix file. In addition, the controller hostname will be added as additional information (currently it's encoded as part of the pathname of the tarball).

3. This PR will change the archive directory structure to make it deeper and not as wide, by using the MD5 string of the tarball to implement an (on-disk) trie (like git, or gluster): e.g. the MD5 sum 7341243af6d9af1025cbec71023dbff7 will generate a pathname of 73/41/7341243af6d9af1025cbec71023dbff7 to a directory that will hold the unpacked "container" tarball.

### CODE REVIEW

**Comment by tenstormavi** [comment]
[`pbench-move-results` NG and its support structures on the server side](https://trello.com/c/2igA5oks/41-pbench-move-results-ng-and-its-support-structures-on-the-server-side)

File: server/pbench/bin/pbench-inotify [hunk 1]
```diff
@@ -176,17 +176,22 @@ class EventHandler(pyinotify.ProcessEvent):
...
-            print("\nTRIGERRED BY: %s" % filename)
-        script_to_run = associations[filename].script
-        args = associations[filename].args
-        list = open(associations[filename].list_file, 'r')
-        fcntl.flock(list, fcntl.LOCK_EX)
-        new_links = list.readlines()
-        fcntl.flock(list, fcntl.LOCK_UN)
-        list.close()
-        associations[filename].compare_and_dispatch(new_links)
+        if event.path == "/pbench/archive/fs-version-001":
```
Comment (ndokos) [inline server/pbench/bin/pbench-inotify]: No hardwired  paths please (and 001 is wrong). If we need the complete path, we should get it from the config file. If not we should match on as little as possible (perhaps "fse-version-002", or even "002"). Even that should come from the config file. Do we need to match on any version *greater or equal* to 002?

File: server/pbench/bin/pbench-inotify [hunk 2]
```diff
@@ -219,6 +224,10 @@ def main():
     # initialization period: any pbench-move-results will have to wait
     # for this to finish.
 
+    wd1=wm.add_watch("/pbench/archive/fs-version-001", mask)
```
Comment (ndokos) [inline server/pbench/bin/pbench-inotify]: This has to agree with the path that we give to `pbench-move-results` (iow, what we put in the `pbench-results-host-info` file). We should specify this path in the config file and retrieve it from there, both in this script and also in the iinstallation script that creates that file.

File: server/pbench/bin/pbench-inotify [hunk 3]
```diff
@@ -219,6 +224,10 @@ def main():
     # initialization period: any pbench-move-results will have to wait
     # for this to finish.
```
Comment (ndokos) [inline server/pbench/bin/pbench-inotify]: What does this mean? `pbench-move-results` are run asynchronously by agents all over the place. They might very well be depositing tarballs into this directory even as this script initializes. But IIRC, pbench-move-results takes a lock on the directory, so we can take the same lock exclusively: that way we wait for them to finish, before starting our initialization. But then we have to process any tarballs that we find in this directory, something which I don't see in the code currently.

**Comment by tenstormavi** [comment]
@ndokos @portante PR is updated with the new changes PTAL.

File: server/pbench/bin/pbench-inotify [hunk 4]
```diff
@@ -168,6 +179,26 @@ def cleanit():
...
+    """
+    threading.Timer(60.0, run_shim).start()  # time in seconds
+    subprocess.Popen("pbench-server-prep-version-002")
+
+def sync_satellite():
+    """
+    Thread that runs periodically to sync the results from satellite host.
+    """
+    threading.Timer(60.0, sync_satellite).start()  # time in seconds
+    mutex = threading.Lock()
```
Comment (ndokos) [inline server/pbench/bin/pbench-inotify]: This is a private mutex: it cannot work by definition.

**Comment by tenstormavi** [comment]
@ndokos @portante Changes made as requested. PTAL

File: server/pbench/bin/pbench-inotify [hunk 5]
```diff
@@ -104,15 +103,15 @@ def sync_satellite():
...
+    with open(lock_file) as f:
+        fcntl.flock(f, fcntl.LOCK_EX)
     prefix = conf.get('pbench-sync', 'prefix')
     satellite_host = conf.get('pbench-sync', 'satellite_host')
     remotearchive = conf.get('pbench-inotify', 'archive')
     subprocess.Popen(["pbench-sync-satellite-inotify", str(prefix), str(satellite_host), str(remotearchive)])
-    fcntl.flock(list, fcntl.LOCK_UN)
-    list.close()
+    with open(lock_file) as f:
+        fcntl.flock(f, fcntl.LOCK_UN)
```
Comment (ndokos) [inline server/pbench/bin/pbench-inotify]: with open(...) as f:
         fcntl.flock(f....)
         prefix = ...
         ...
         subprocess.Popen(...)
         fcntl.flock(f, fcntl.LOCK_UN)

I think.

Comment (tenstormavi) [reply server/pbench/bin/pbench-inotify]: Right.

**Comment by ndokos** [comment]
@portante: please review!

**Comment by portante** [comment]
Thanks Avinash, reviewing your code now.  Can you rebase this to latest `master` in the meantime?

**Comment by portante** [comment]
@tenstormavi, please see https://github.com/distributed-system-analysis/pbench/pull/848, and make sure those files don't come back as a result of your changes, but show up as changes to the original non-`inotify` files.   Thanks!

**Comment by tenstormavi** [comment]
@portante @ndokos Made the changes as requested PTAL.

Changes Includes:
1. rebase with latest `master`
2. squashed all 11 commits into one
3. proper commit comment
4. Remove the two copies of a file
5. Included the example of serviced file and conf file to remove cron operations.

**Comment by tenstormavi** [comment]
@ndokos @portante rebased with the latest master. Update the unit tests also. PTAL

**Comment by portante** [comment]
See #797 closing comment.

**Pull request closed without merging**